### PR TITLE
Upgrade mkldnn to version 0.18.1

### PR DESCRIPTION
**Pull request by gujinghui**
Upgrade mkldnn to version 0.18.1
Fix the MKLDNN build issue if linking with MKL 2019.

### CODE REVIEW

**Comment by gujinghui** [comment]
@yinghai 
Pls help review.
This PR only upgrade MKLDNN to 0.18.1 to fix the build issue when linking with MKL2019.
No other changes in mkldnn-bridge.

**Comment by gujinghui** [comment]
@jgong5  @Jianhui-Li

**Comment by uyongw** [comment]
With the mkl-dnn upgraded to v0.18.1, it will also fix the correctness issue https://github.com/pytorch/pytorch/issues/18218
More details can be found from https://github.com/intel/mkl-dnn/issues/431

**Comment by Jianhui-Li** [comment]
This PR address MKL-DNN build issue in https://github.com/pytorch/pytorch/issues/18218

**Comment by SsnL** [comment]
@uyongw Thanks for fixing that issue! What is the specific configuration that caused the buggy behavior? How do you suggest that we could add a test in pytorch to prevent regression?

**Comment by uyongw** [comment]
> @uyongw Thanks for fixing that issue! What is the specific configuration that caused the buggy behavior? How do you suggest that we could add a test in pytorch to prevent regression?

This means to add group convolution with (input-channel / group) < 16 (or 8). For example:
group=64, input-channel=256, output-channel=1024, input h/w=28, output h/w=14, kernel h/w=3, stride=2, padding=1

**Comment by fracting** [comment]
@uyongw @SsnL 

```
import torch
from torch import nn

in_channel = 512
out_channel = 1024
groups = 64
in_channel_per_group = in_channel // groups
out_channel_per_group = out_channel // groups
conv2d = nn.Conv2d(in_channel, out_channel, kernel_size=(3,3), stride=(2,2), padding=(1,1), groups=groups, bias=False)

input_tensor = torch.randn(1, 512, 28, 28)
output_tensor = conv2d(input_tensor)

output_tensor_slices = []
for i in range(groups):
  conv2d_1g = nn.Conv2d(in_channel_per_group, out_channel_per_group, kernel_size=(3,3), stride=(2,2), padding=(1,1), groups=1, bias=False)
  conv2d_1g.weight.data = conv2d.weight.data[i*out_channel_per_group:(i+1)*out_channel_per_group,:,:,:]
  input_tensor_slice = input_tensor[:,i*in_channel_per_group:(i+1)*in_channel_per_group,:,:]
  output_tensor_slice = conv2d_1g(input_tensor_slice)
  output_tensor_slices.append(output_tensor_slice)

output_cat = torch.cat(output_tensor_slices, 1)
diff = output_tensor - output_cat
print(diff)
# assertAlmostEqual(output_tensor, output_cat)
```

The above code demonstrates the bug well for me on Xeon, it also validates the fix. If you like I can clean up and submit a pull request adding a test to `test/test_nn.py` for you to review.

Even with the pytorch test, I guess it would be nice to have a C++ test in mkl-dnn repository itself, since mkl-dnn is used by multiple downstream projects. What do you think?

**Comment by fracting** [comment]
@SsnL @uyongw 

After reading a bit more inside `test_nn.py`, I was wondering why `_test_Conv2d_naive_groups` and `test_Conv2d_groups_nobias` didn't expose the bug, since the hyperparameters used in the tests matches the condition pointed out by @uyongw `(input-channel / group) < 16 (or 8)`

Then I decided to print the output of the test case, it turns out that on my `i7` machine, `output - torch.cat([output1, output2], 1)` is exactly zero, however, on my `xeon` machine, `output - torch.cat([output1, output2], 1)` is slightly away from zero while not too far away, for example:

```
tensor([[[[ 5.9605e-08,  2.9802e-08, -1.1921e-07,  1.1921e-07],
          [ 2.2352e-08,  3.7253e-08,  0.0000e+00,  0.0000e+00],
          [ 0.0000e+00, -1.4901e-08,  0.0000e+00,  5.9605e-08],
          [ 0.0000e+00,  1.4901e-08,  0.0000e+00,  0.0000e+00]],

         [[-2.9802e-08,  0.0000e+00, -5.9605e-08,  1.1921e-07],
          [ 0.0000e+00,  1.4901e-08,  2.9802e-08,  0.0000e+00],
          [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  0.0000e+00],
          [ 1.4901e-08, -4.4703e-08,  1.4901e-08, -4.0978e-08]],

         [[-5.9605e-08,  0.0000e+00,  0.0000e+00, -5.9605e-08],
          [ 5.9605e-08,  0.0000e+00,  0.0000e+00,  0.0000e+00],
          [ 7.4506e-08,  0.0000e+00,  0.0000e+00, -4.4703e-08],
          [ 5.9605e-08,  0.0000e+00,  2.9802e-08,  0.0000e+00]],

         [[ 1.4901e-08,  1.4901e-08, -2.9802e-08, -1.7881e-07],
          [ 8.9407e-08,  0.0000e+00,  0.0000e+00,  0.0000e+00],
          [ 2.9802e-08, -1.4901e-08, -5.9605e-08,  5.9605e-08],
          [ 0.0000e+00,  2.9802e-08,  0.0000e+00,  5.9605e-08]]],


        [[[-2.9802e-08,  2.9802e-08,  0.0000e+00,  5.9605e-08],
          [-5.9605e-08, -2.9802e-08, -2.9802e-08, -2.9802e-08],
          [ 2.9802e-08,  1.4901e-08,  0.0000e+00, -1.1921e-07],
          [ 0.0000e+00,  1.4901e-08, -5.9605e-08,  0.0000e+00]],

         [[-5.9605e-08,  2.9802e-08,  0.0000e+00, -2.9802e-08],
          [ 0.0000e+00,  0.0000e+00,  0.0000e+00,  4.4703e-08],
          [-2.9802e-08,  0.0000e+00,  5.9605e-08,  0.0000e+00],
          [ 1.1921e-07,  5.9605e-08,  0.0000e+00, -2.9802e-08]],

         [[ 5.9605e-08, -1.4901e-08, -2.9802e-08,  0.0000e+00],
          [-7.4506e-08, -1.1921e-07,  5.9605e-08,  1.4901e-08],
          [ 0.0000e+00,  1.1921e-07, -2.9802e-08,  2.9802e-08],
          [-5.9605e-08,  0.0000e+00,  0.0000e+00,  2.9802e-08]],

         [[-5.9605e-08, -5.9605e-08,  0.0000e+00, -5.9605e-08],
          [-5.9605e-08,  2.9802e-08, -1.1921e-07, -5.9605e-08],
          [ 0.0000e+00, -2.9802e-08,  0.0000e+00,  0.0000e+00],
          [-5.9605e-08, -5.9605e-08, -1.4901e-08,  2.9802e-08]]]],
       dtype=torch.float32, grad_fn=<SubBackward0>)
```
```
# test/common_utils.py
285 class TestCase(expecttest.TestCase):
286     precision = 1e-5
```
Since the default `prec` defined in `common_utils.py` is `1e-05`, the above small difference was not detected by the test case.

Decreasing precision from `1e-05` to `1e-07` exposes this bug in 10/10 runs. (Even setting precision to `1e-06` failed to expose this bug in 10/10 runs)

However, I'm not sure if this would cause false alarm for other tests.

**Comment by uyongw** [comment]
@fracting I forgot to mention that to reproduce the issue the output-channel should be multiple of 16 (in additional to group > 1 && input-channel / group) < 16.

The _test_Conv2d_naive_groups result is a numeric accuracy issue. IMHO, both i7/xeon results are 'correct'. However your original case is a correctness issue.

**Comment by fracting** [comment]
Thanks @uyongw !

I can confirm the below test case exposes the bug even without changing precision.
In the below case, on xeon machine, the first half of `output` is consistent with `output1`, while the second half of `output` is totally different from `output2`. This seems to match what https://github.com/intel/mkl-dnn/commit/c502e5f45410d9b2bb2155cb99364eaf2360c6db trying to fix.

```
    def test_Conv2d_groups_nobias_large_output_channel(self):
        dev_dtypes = [("cpu", torch.float)]
        if TEST_CUDA:
            dev_dtypes += [("cuda", torch.float), ("cuda", torch.half)]
        for device, dtype in dev_dtypes:
            m = nn.Conv2d(4, 16, kernel_size=3, groups=2, bias=False).to(device, dtype)
            i = torch.randn(2, 4, 6, 6, device=device, dtype=dtype, requires_grad=True)
            output = m(i)
            grad_output = torch.randn(2, 16, 4, 4, device=device, dtype=dtype)
            output.backward(grad_output)

            m1 = nn.Conv2d(2, 8, kernel_size=3, bias=False).to(device, dtype)
            m1.weight.data.copy_(m.weight.data[:8])
            i1 = Variable(i.data[:, :2].contiguous(), requires_grad=True)
            output1 = m1(i1)
            output1.backward(grad_output[:, :8].contiguous())

            m2 = nn.Conv2d(2, 8, kernel_size=3, bias=False).to(device, dtype)
            m2.weight.data.copy_(m.weight.data[8:])
            i2 = Variable(i.data[:, 2:].contiguous(), requires_grad=True)
            output2 = m2(i2)
            output2.backward(grad_output[:, 8:].contiguous())

            self.assertEqual(output, torch.cat([output1, output2], 1))
            self.assertEqual(i.grad.data,
                             torch.cat([i1.grad.data, i2.grad.data], 1),
                             dtype2prec[dtype])
            self.assertEqual(m.weight.grad.data,
                             torch.cat([m1.weight.grad.data, m2.weight.grad.data], 0),
                             1e-1 if dtype == torch.half else dtype2prec[dtype])
```

**Comment by uyongw** [comment]
@fracting Thank you very much! This is the case we need indeed. @SsnL

**Comment by fracting** [comment]
@uyongw @SsnL I created https://github.com/pytorch/pytorch/pull/18519 for you to review, thank you!

**Comment by SsnL** [comment]
@uyongw @fracting Thank you all!
